src/main/java/org/folio/circulation/resources/CirculationRulesResource.java

```diff
@@ -115,14 +147,78 @@ private void put(RoutingContext routingContext) {
+    clients.circulationRulesStorage().put(rulesInput.copy())
+      .thenAccept(res -> res.applySideEffect(response -> {
         if (response.getStatusCode() == 204) {
           SuccessResponse.noContent(routingContext.response());
         } else {
           ForwardResponse.forward(routingContext.response(), response);
         }
-    }, cause -> cause.writeTo(routingContext.response())));
+      }, cause -> cause.writeTo(routingContext.response())));
+  }
+
+   private boolean policyExists(Map<String, Set<String>> existingPolicyIds,
+                                String policyType, List<CirculationRulesParser.PolicyContext> policies) {
+
+    return MapUtils.isNotEmpty(existingPolicyIds) && isNotEmpty(existingPolicyIds.get(policyType))
+        && existingPolicyIds.get(policyType).contains(
+          policies.get(FIRST_ELEMENT_OF_LIST)
+            .getChild(POLICY_ID_POSITION_NUMBER).getText());
+  }
+
+  private CompletableFuture<Result<Map<String, Set<String>>>> getExistingPolicyIds(Clients clients) {
+
+    CollectionResourceClient loanPolicyClient = clients.loanPoliciesStorage();
+    CollectionResourceClient noticePolicyClient = clients.patronNoticePolicesStorageClient();
+    CollectionResourceClient requestPolicyClient = clients.requestPoliciesStorage();
+    CollectionResourceClient overdueFinePolicyClient = clients.overdueFinesPoliciesStorage();
+    CollectionResourceClient lostItemFeePolicyClient = clients.lostItemPoliciesStorage();
+    Map<String, Set<String>> ids = new HashMap<>();
+
+    return Result.ofAsync(() -> ids)
+      .thenCombineAsync(getExistingPolicyIds(loanPolicyClient, "loanPolicies", "l"),
+        (resultTotalIds, resultNewIds) -> combine(resultTotalIds, resultNewIds,
+          this::getTotalMap))
+      .thenCombineAsync(getExistingPolicyIds(noticePolicyClient, "patronNoticePolicies", "n"),
+        (resultTotalIds, resultNewIds) -> combine(resultTotalIds, resultNewIds,
+          this::getTotalMap))
+      .thenCombineAsync(getExistingPolicyIds(requestPolicyClient, "requestPolicies", "r"),
+        (resultTotalIds, resultNewIds) -> combine(resultTotalIds, resultNewIds,
+          this::getTotalMap))
+      .thenCombineAsync(getExistingPolicyIds(overdueFinePolicyClient, "overdueFinePolicies", "o"),
+        (resultTotalIds, resultNewIds) -> combine(resultTotalIds, resultNewIds,
+          this::getTotalMap))
+      .thenCombineAsync(getExistingPolicyIds(lostItemFeePolicyClient, "lostItemFeePolicies", "i"),
+        (resultTotalIds, resultNewIds) -> combine(resultTotalIds, resultNewIds,
+          this::getTotalMap));
+  }
+
+  private Map<String, Set<String>> getTotalMap(
+    Map<String, Set<String>> totalMap, Map<String, Set<String>> newMap) {
+    totalMap.putAll(newMap);
+    return totalMap;
+  }
+
+  private CompletableFuture<Result<Map<String, Set<String>>>> getExistingPolicyIds(
+    CollectionResourceClient client, String entityName, String policyType) {
```

Comment:
Is the entity name intended to be the type of policy? If so, please could this be renamed to reflect that more clearly

Reply:
renamed

Reply:
Thank you